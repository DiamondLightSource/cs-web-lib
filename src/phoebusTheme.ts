import { createTheme } from "@mui/material/styles";

export const phoebusTheme = createTheme({
  palette: {
    primary: {
      main: "#D2D2D2",
      // light: currently calculated automatically by MUI
      // dark: currently calculated automatically by MUI
      contrastText: "#000000"
    }
  },
  typography: {
    fontFamily: "Liberation Sans",
    fontSize: 14,
    button: {
      textTransform: "none"
    }
  }
});
