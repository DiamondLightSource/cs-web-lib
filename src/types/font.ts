import { CSSProperties } from "react";

// assigned int values starting from zero
// also functions as a type
export enum FontStyle {
  Regular,
  Bold,
  Italic,
  BoldItalic
}

export class Font {
  // If no number is provided inherit from CSS.
  private size?: number;
  private style: FontStyle;
  private typeface: string;
  // TODO: name is currently unused
  private name?: string;

  public constructor(
    size?: number,
    style?: FontStyle,
    typeface?: string,
    name?: string
  ) {
    this.typeface = typeface ?? "Liberation sans";
    this.style = style ?? FontStyle.Regular;
    this.size = size;
    this.name = name;
  }

  /**
   * Returns a CSSProperties object containing all parameters defined by the
   * user and default properties for optional parameters not input by the user
   */
  public css(): CSSProperties {
    // Convert from px to REM by dividing pixels by default browser font size 16px
    // If a user has manually adjusted their browser font size, this might not be scaled
    // correctly, but given bob screens are pixel placed we shouldn't expect this to scale
    const fontSize = this.size ? `${this.size / 16}rem` : undefined;
    const fontWeight =
      this.style === FontStyle.Bold || this.style === FontStyle.BoldItalic
        ? "bold"
        : "normal";
    const fontStyle =
      this.style === FontStyle.Italic || this.style === FontStyle.BoldItalic
        ? "italic"
        : "normal";
    const style: CSSProperties = {
      // Fall back to sans-serif.
      fontFamily: `${this.typeface},sans-serif`,
      fontWeight: fontWeight,
      fontStyle: fontStyle
    };
    if (fontSize) {
      style.fontSize = fontSize;
    }
    return style;
  }
}
