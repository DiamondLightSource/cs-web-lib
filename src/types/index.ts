export { AbsolutePosition } from "./position";
export { RelativePosition } from "./position";
export { Color } from "./color";
export { PV } from "./pv";
export { Font, FontStyle } from "./font";
export { Border, BorderStyle } from "./border";
