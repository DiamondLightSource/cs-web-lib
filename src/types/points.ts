export class Point {
  public x: number;
  public y: number;

  public constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}

export class Points {
  public values: Array<Point>;

  public constructor(values: Array<Point>) {
    this.values = values;
  }
}
