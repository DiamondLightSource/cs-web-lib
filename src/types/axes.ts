import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx;
    this.autoScale = true;
    this.autoScaleThreshold = 0.95;
    this.axisColor = new Color("rgb(0, 0, 0");
    this.axisTitle = "";
    this.showGrid = false;
    this.gridColor = new Color("rgb(0, 0, 0");
    this.dashGridLine = true;
    this.timeFormat = 0;
    this.scaleFormat = "";
    this.scaleFont = new Font();
    this.titleFont = new Font(10, FontStyle.Bold);
    this.visible = true;
    this.logScale = false;
    this.leftBottomSide = true;
    this.minimum = 0;
    this.maximum = 100;
    // If axis has index 1 or above, assume y axis
    this.index > 0 ? (this.yAxis = true) : (this.yAxis = false);
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
