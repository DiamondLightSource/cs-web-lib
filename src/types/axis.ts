import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public color: Color;
  public title: string;
  public showGrid: boolean;
  public visible: boolean;
  public logScale: boolean;
  public autoscale: boolean;
  public maximum: number;
  public minimum: number;
  public titleFont: Font;
  public scaleFont: Font;
  public onRight: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor({
    xAxis = false,
    color = new Color("rgb(0, 0, 0"),
    title = "",
    showGrid = false,
    visible = true,
    logScale = false,
    autoscale = false,
    minimum = 0,
    maximum = 100,
    titleFont = new Font(FontStyle.Bold),
    scaleFont = new Font(),
    onRight = false
  } = {}) {
    this.color = color;
    this.title = title || (xAxis ? "X" : "Y");
    this.showGrid = showGrid;
    this.visible = visible;
    this.logScale = logScale;
    this.autoscale = autoscale;
    this.minimum = minimum;
    this.maximum = maximum;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.onRight = onRight;
  }
}

export type Axes = Axis[];
