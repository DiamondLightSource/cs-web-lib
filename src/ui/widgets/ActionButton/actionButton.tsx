import React, { useContext } from "react";
import { WidgetActions, executeActions } from "../widgetActions";
import { Widget } from "../widget";
import { PVComponent, PVWidgetPropType } from "../widgetProps";
import classes from "./actionButton.module.css";
import { registerWidget } from "../register";
import {
  ActionsPropType,
  StringPropOpt,
  InferWidgetProps,
  ColorPropOpt,
  FontPropOpt,
  BorderPropOpt,
  BoolPropOpt
} from "../propTypes";
import { Color } from "../../../types/color";
import { Font } from "../../../types/font";
import { Border } from "../../../types/border";
import { MacroContext } from "../../../types/macros";
import { ExitFileContext, FileContext } from "../../../misc/fileContext";
import { Button, ThemeProvider } from "@mui/material";

import { defaultColours } from "../../../colourscheme";

export interface ActionButtonProps {
  text: string;
  disabled?: boolean;
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
  image?: string;
  backgroundColor?: Color;
  foregroundColor?: Color;
  border?: Border;
  font?: Font;
  actions?: WidgetActions;
  visible?: boolean;
}

const ActionButtonPropType = {
  text: StringPropOpt,
  actions: ActionsPropType,
  image: StringPropOpt,
  backgroundColor: ColorPropOpt,
  foregroundColor: ColorPropOpt,
  font: FontPropOpt,
  border: BorderPropOpt,
  visible: BoolPropOpt,
  disabled: BoolPropOpt
};

const ActionButtonWidgetProps = {
  ...ActionButtonPropType,
  ...PVWidgetPropType
};

// Menu button which also knows how to write to a PV
export const ActionButtonWidget = (
  props: InferWidgetProps<typeof ActionButtonWidgetProps> & PVComponent
): JSX.Element => {
  // Function to send the value on to the PV
  const files = useContext(FileContext);
  const exitContext = useContext(ExitFileContext);
  const parentMacros = useContext(MacroContext).macros;
  function onClick(event: React.MouseEvent<HTMLButtonElement>): void {
    if (props.actions !== undefined)
      executeActions(
        props.actions as WidgetActions,
        files,
        exitContext,
        parentMacros
      );
  }
  return (
    <ThemeProvider theme={defaultColours}>
      <Button
        variant="contained"
        disabled={props.disabled}
        sx={{
          height: "100%",
          width: "100%",
          fontFamily: props.font?.css() ?? "",
          color: 
            props.foregroundColor?.toString() ??
            defaultColours.palette.primary.contrastText,
          backgroundColor: 
            props.backgroundColor?.toString() ??
            defaultColours.palette.primary.main,
          border: props.border?.css() ?? ""
        }}
        onClick={onClick}
      >
        {props.image !== undefined ? (
          <figure className={classes.figure}>
            <img
              style={{ width: "100%", display: "block" }}
              src={props.image}
              alt={props.image}
            ></img>
            <figcaption>{props.text}</figcaption>
          </figure>
        ) : (
          (props.text ?? "")
        )}
      </Button>
    </ThemeProvider>
  );
};

export const ActionButton = (
  props: InferWidgetProps<typeof ActionButtonWidgetProps>
): JSX.Element => <Widget baseWidget={ActionButtonWidget} {...props} />;

registerWidget(ActionButton, ActionButtonWidgetProps, "actionbutton");
