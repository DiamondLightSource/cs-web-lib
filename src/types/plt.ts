import { Axes, Axis } from "./axis";
import { Color } from "./color";
import { Font } from "./font";
import { Trace } from "./trace";

export class Plt {
  public title: string;
  public axes: Axes;
  public pvlist: Trace[];
  public backgroundColor: Color;
  public foregroundColor: Color;
  public scroll: boolean;
  public scrollStep: number;
  public updatePeriod: number;
  public start: string;
  public end: string;
  public showGrid: boolean;
  public titleFont: Font;
  public scaleFont: Font;
  public labelFont: Font;
  public legendFont: Font;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property.
   */
  public constructor({
    title = "",
    axes = [new Axis()],
    pvlist = [new Trace()],
    background = Color.WHITE,
    foreground = Color.BLACK,
    scroll = true,
    grid = false,
    scrollStep = 5,
    updatePeriod = 1,
    titleFont = new Font(),
    labelFont = new Font(),
    legendFont = new Font(),
    scaleFont = new Font(),
    start = "1 minute",
    end = "now"
  } = {}) {
    this.backgroundColor = background;
    this.foregroundColor = foreground;
    this.title = title;
    this.scroll = scroll;
    this.scrollStep = scrollStep;
    this.axes = axes;
    this.pvlist = pvlist;
    this.updatePeriod = updatePeriod;
    this.showGrid = grid;
    this.start = start;
    this.end = end;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.labelFont = labelFont;
    this.legendFont = legendFont;
  }
}
