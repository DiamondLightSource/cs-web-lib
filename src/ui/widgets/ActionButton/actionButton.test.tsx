import React from "react";
import { ActionButtonComponent } from "./actionButton";
import renderer from "react-test-renderer";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { vi } from "vitest";
import { contextRender } from "../../../testResources";
import { ThemeProvider } from "@mui/material";
import { phoebusTheme } from "../../../phoebusTheme";
import { Color } from "../../../types";

// Pass the theme in with the component to access default values
const mock = vi.fn();
const actionButton = (props: any) => (
  <ThemeProvider theme={phoebusTheme}>
    <ActionButtonComponent text={"hello"} onClick={mock} {...props} />
  </ThemeProvider>
);

describe("<ActionButton />", (): void => {
  test("it matches the snapshot", (): void => {
    const snapshot = renderer.create(actionButton({}));
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test("it renders a button with default style from theme", (): void => {
    const { getByRole } = contextRender(actionButton({}));
    const button = getByRole("button");
    expect(button).toHaveStyle({
      "background-color": "rgb(210, 210, 210)",
      color: "rgb(0, 0, 0)",
      width: "100px",
      height: "30px"
    });
    // For some reason, background colour doesn't like to be passed as an object so pass as string
    expect(button.textContent).toEqual("hello");
  });

  test("it renders a button with style from props", (): void => {
    const { getByRole } = render(
      actionButton({
        foregroundColor: Color.fromRgba(155, 160, 209),
        backgroundColor: Color.fromRgba(10, 200, 1),
        height: 40,
        width: 60
      })
    );
    const button = getByRole("button");
    expect(button).toHaveStyle({
      "background-color": "rgb(10, 200, 1)",
      color: "rgb(155, 160, 209)",
      width: "60px",
      height: "40px"
    });
    // For some reason, background colour doesn't like to be passed as an object so pass as string
  });

  test("button variation changes if transparent", (): void => {
    const { getByRole } = render(actionButton({ transparent: true }));
    const button = getByRole("button");
    expect(button).toHaveStyle("background-color: rgba(0, 0, 0, 0)");
  });

  test("button is rotated correctly when rotationStep is set", async () => {
    const { getByRole } = render(
      actionButton({ rotationStep: 1, height: 40, width: 60 })
    );
    const button = await getByRole("button");
    expect(button).toHaveStyle({
      transform: "rotate(-90deg) translateY(10px) translateX(10px)",
      height: "60px",
      width: "40px"
    });
  });

  test("whether button is disabled", async () => {
    const { getByRole } = render(actionButton({ enabled: false }));
    const button = await getByRole("button");
    expect(button).toBeDisabled();
  });

  test("function called on click", async (): Promise<void> => {
    const { getByRole } = render(actionButton({}));
    const button = getByRole("button") as HTMLButtonElement;
    fireEvent.click(button);
    await waitFor(() => expect(mock).toHaveBeenCalled());
  });
});
