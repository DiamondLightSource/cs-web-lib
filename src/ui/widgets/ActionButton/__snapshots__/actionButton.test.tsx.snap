// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ActionButton /> > it matches the snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-q0v14n-MuiButtonBase-root-MuiButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[MockFunction spy]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    style={
      {
        "display": "block",
        "lineHeight": 1,
        "transform": "rotate(0deg)",
      }
    }
  >
    hello
  </span>
</button>
`;
